### ISSUE TITLE
Update transformer based on opennmt-tf

### ISSUE THREAD

**Issue by helson73** [OpenNMT/OpenNMT-py#432]
1. forward() in  Transformer is not compatible with Translator.py (line 153), which requires context_lengths.
2. Transformer actually doesn't converge using default options. (PPL is just ridiculous)
3. Too many parameters are reported. 

Some of them are already reported in other threads.
It could be hardly used in current state.

**Comment by JianyuZhan** [comment]
1. This is a bug due to a recent check-in:https://github.com/OpenNMT/OpenNMT-py/commit/19dc63b3db77ff397638776e0b63c6de08140d59 , will fix !

2 & 3.  Could you post your test result (including your parameter settings)  for us to reproduce and investigate further?  If you have any clue, also welcome feeback!

Thanks.

**Comment by helson73** [comment]
If I have an option, I will use official release (Tensor2Tensor), but for some reasons, I have to use pytorch. 
I will continue to do further hack on this, if I got further progress, I will post on this thread.

**Comment by helson73** [comment]
In the original paper, target and output embeddings are shared, and mine is not. Maybe this made a difference.

**Comment by JianyuZhan** [comment]
Thanks, will fix  problem 1 today. And will investigate later.

**Comment by davidecaroselli** [comment]
@helson73 same problem here! Perplexity is ridiculously high. I tried adding `-share_decoder_embeddings` (I think this is the option you were referring to), but nothing changed. Here's the command I'm running:
```
python train.py -data mydata/data -save_model mydata/model -gpuid 0 -epochs 60 \
    -learning_rate_decay 0.98 -batch_size 100 -warmup_steps 32000 \
    -encoder_type transformer -decoder_type transformer \
    -word_vec_size 512 -rnn_size 512 -layers 2 -share_decoder_embeddings
```

I'm running also a classic `rnn` network for reference:
```
python train.py -data mydata/data -save_model mydata/model -gpuid 0 -epochs 50 \
    -learning_rate_decay 0.9 -batch_size 128 -warmup_steps 16000 \
    -encoder_type rnn  -decoder_type rnn \
    -word_vec_size 512 -rnn_size 512 -layers 2
```

After 12 epochs reached `Validation perplexity: 3.53089` while the transformer one is stuck at `Validation perplexity: 745.375`

**Comment by sebastianGehrmann** [comment]
The transformer module is not fully documented yet, but it should work with the correct set of parameters. Try the following (in addition to batch_size, gpuid, data etc.) : 

`-layers  4 -rnn_size 1024  -word_vec_size 1024 -epochs 50  -max_grad_norm 0 -optim adam -encoder_type transformer -decoder_type transformer -position_encoding -dropout 0.2 -param_init 0 -warmup_steps 2000 -learning_rate 0.05 -decay_method noam`

You might have to play around with the learning rate a bit, but it should converge within 40 epochs (for me, typically around epoch 37 on a small data set)

**Comment by guillaumekln** [comment]
The recommended parameters should be as close as possible to the [`transformer_base`](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py#L791) configuration used in Tensor2Tensor. It's likely that these hyperparameters were validated at Google scale.

It should also be pointed out that this "base" configuration slightly differs from the paper. In particular, the "Add & Norm" layer is split and layer normalization is moved to the block input. Also dropout is applied after the ReLU activation. It might be a good idea to apply these changes here as well.

**Comment by davidecaroselli** [comment]
@sebastianGehrmann much better now! At epoch 13 the engine reached `Validation perplexity: 7.79672` and it is decreasing.

Anyway, do you confirm that the number of parameters is too high respect to the same topology with RNN? I have ~180M parameters, almost the same amount of the RNN network.

**Comment by srush** [comment]
thanks @guillaumekln . Were you able to replicate results in opennmt-tf? I will try to make our model as close as possible to your version, and then publish the exact code.

**Comment by guillaumekln** [comment]
The implementation over at OpenNMT-tf is very close to the one implemented in Tensor2Tensor and so gets competitive results. However, batch size is important when using the Noam decay and they use a different semantic for it: the number of tokens. This is also something to consider and explore.

Here are the direct links to the implementation if you want to look it up:

* https://github.com/OpenNMT/OpenNMT-tf/blob/master/opennmt/utils/transformer.py
* https://github.com/OpenNMT/OpenNMT-tf/blob/master/opennmt/encoders/self_attention_encoder.py
* https://github.com/OpenNMT/OpenNMT-tf/blob/master/opennmt/decoders/self_attention_decoder.py

**Comment by helson73** [comment]
Other issues about gaps between transformer-py and T2T:
1) Label smoothing. [https://github.com/tensorflow/tensor2tensor/blob/5c8009561c4604be53e27310d0014ce69176c5db/tensor2tensor/layers/common_layers.py#L1626](url)
2) Shared embeddings for decoder: When embeddings are shared between target input and output, input embeddings should be multiplied by sqrt(d-model), where d-model is dimension of embeddings. [https://github.com/tensorflow/tensor2tensor/blob/5c8009561c4604be53e27310d0014ce69176c5db/tensor2tensor/layers/modalities.py#L112](url)

**Comment by helson73** [comment]
@davidecaroselli In the paper, #parameters per hyperparameters are reported in detail. 
As far as I found, if input and output embeddings of decoder are shared, number of parameters is quite close to the one reported.

**Comment by helson73** [comment]
Seems like embedding weight tying in Transformer is a little bit more complicated than I expected. 
In [Attention is All You Need](https://arxiv.org/pdf/1706.03762.pdf) : 
> We also use the usual learned linear transformation
and softmax function to convert the decoder output to predicted next-token probabilities. In
our model, we share the same weight matrix between **the two embedding layers and the pre-softmax
linear transformation**, similar to [29](https://arxiv.org/pdf/1608.05859.pdf).

I have to check the sourcecode further, but it sounds like they applied Three-Way-Weight-Tying (shared embeddings for source, target input, target output).

**Comment by helson73** [comment]
@srush 
I applied following changes as @guillaumekln suggested, 
1) Move norm layer to the block input,
2) Apply dropout after ReLU in point-wise feed forward block,
and trained on small dataset (WAT 2016 Japanese-to-English, randomly selected 100K sentence pairs) for 20 epochs,
got following result.
Before:
> 
Train perplexity: 32.4537
Train accuracy: 38.7784
Validation perplexity: 56.2081
Validation accuracy: 33.0021

After:
> 
Train perplexity: 29.8781
Train accuracy: 40.5064
Validation perplexity: 52.0337
Validation accuracy: 34.0583

**Comment by guillaumekln** [comment]
I forgot to mention that layer normalization is also applied on the encoder and decoder final output.

Also, they did not always use weight sharing. For example, the big ENFR model was trained without it.

**Comment by helson73** [comment]
@guillaumekln I also applied that, actually I did those changes based on opennmt-tf sourcecode.
About weight sharing, once it was switched on, PPL became extremely high, I am trying to figure out what causes this.

**Comment by srush** [comment]
1) We should have embedding weight sharing implemented. `share_embeddings` and `share_decoder_embeddings`. Do you see issues with those?

2) Label smoothing shouldn't be so bad. I think we could just replace this line 
https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Loss.py#L111

**Comment by srush** [comment]
@helson73 thanks for the PR.

Do you want to try to do label smoothing as well? Here's roughly how I would do it: 

```python
class LabelSmooth:
  def __init__(self, confidence, vocab_size):
     vocab_size = float(vocab_size)
     self.crit = nn.KLDivLoss()
     self.low_confidence = (1.0 - confidence) / (vocab_size - 1)
     self.confidence = confidence
     entropy = \
        confidence * torch.log(confidence) + (vocab_size - 1) *
        low_confidence * torch.log(low_confidence + 1e-20)

  def forward(self, log_probs, targets):    
    soft_targets.fill_(low_confidence)
    soft_targets.scatter_(1, targets, confidence)
    return self.crit(log_probs, soft_targets) - entropy
```

**Comment by helson73** [comment]
@srush I implemented as follows, similar with yours, but entropy term is omitted due to it has no influence to learning. 
`class NMTLossCompute(LossComputeBase):
    """
    Standard NMT Loss Computation.
    """
    def __init__(self, generator, tgt_vocab, label_smoothing=0.0):
        super(NMTLossCompute, self).__init__(generator, tgt_vocab)

        weight = torch.ones(len(tgt_vocab))
        weight[self.padding_idx] = 0
        if label_smoothing > 0:
            self.criterion = nn.KLDivLoss(size_average=False)
            self.p_hot = 1 - label_smoothing * (1 - 1 / len(tgt_vocab))
            self.one_hot = torch.randn(1, len(tgt_vocab))
            self.one_hot.fill_(label_smoothing / len(tgt_vocab))
            self.register_buffer('one_hot_vec', self.one_hot)
        else:
            self.criterion = nn.NLLLoss(weight, size_average=False)
        self.label_smoothing = label_smoothing

    def make_shard_state(self, batch, output, range_, attns=None):
        """ See base class for args description. """
        return {
            "output": output,
            "target": batch.tgt[range_[0] + 1: range_[1]],
        }

    def compute_loss(self, batch, output, target):
        """ See base class for args description. """
        scores = self.generator(self.bottle(output))

        target_feed = target.view(-1)
        if self.label_smoothing > 0:
            target_ = Variable(self.one_hot.repeat(target_feed.size(0), 1),
                               requires_grad=False)
            target_.scatter_(1, target_feed.unsqueeze(1), self.p_hot)
            target_feed = target_

        loss = self.criterion(scores, target_feed)
        loss_data = loss.data.clone()

        stats = self.stats(loss_data, scores.data, target.view(-1).data)

        return loss, stats

    def cuda(self):
        super(NMTLossCompute, self).cuda()
        if self.label_smoothing > 0:
            self.one_hot = self.one_hot.cuda()

    def cpu(self):
        super(NMTLossCompute, self).cpu()
        if self.label_smoothing > 0:
            self.one_hot = self.one_hot.cpu()`
Accuray is slightly improved on my small dataset, need further explore.

**Comment by helson73** [comment]
@srush 
About weight sharing embeddings, I didn't see any issue from code. 
But it's quite weird that once I turn weight sharing option on, PPL became extremely high. (only with Transformer, not sure whether same happens to RNN models)

**Comment by JianyuZhan** [comment]
@helson73, you don't need to write your `cuda()` and `cpu()` implementation, since `*LossCompute` is subclass of `nn.Module`, the `self.ont_hot` will get to gpu if the whole object do so.

**Comment by helson73** [comment]
@JianyuZhan Yeah, I see it, actually that's a dirty patch during debugging. 
New code:
```python
class NMTLossCompute(LossComputeBase):
    """
    Standard NMT Loss Computation.
    """
    def __init__(self, generator, tgt_vocab, label_smoothing=0.0):
        super(NMTLossCompute, self).__init__(generator, tgt_vocab)

        weight = torch.ones(len(tgt_vocab))
        weight[self.padding_idx] = 0
        if label_smoothing > 0:
            self.criterion = nn.KLDivLoss(size_average=False)
            self.p_hot = 1 - label_smoothing * (1 - 1 / len(tgt_vocab))
            one_hot = torch.randn(1, len(tgt_vocab))
            one_hot.fill_(label_smoothing / len(tgt_vocab))
            self.register_buffer('one_hot', one_hot)
        else:
            self.criterion = nn.NLLLoss(weight, size_average=False)
        self.label_smoothing = label_smoothing

    def make_shard_state(self, batch, output, range_, attns=None):
        """ See base class for args description. """
        return {
            "output": output,
            "target": batch.tgt[range_[0] + 1: range_[1]],
        }

    def compute_loss(self, batch, output, target):
        """ See base class for args description. """
        scores = self.generator(self.bottle(output))

        target_feed = target.view(-1)
        if self.label_smoothing > 0:
            target_ = Variable(self.one_hot.repeat(target_feed.size(0), 1),
                               requires_grad=False)
            target_.scatter_(1, target_feed.unsqueeze(1), self.p_hot)
            target_feed = target_

        loss = self.criterion(scores, target_feed)
        loss_data = loss.data.clone()

        stats = self.stats(loss_data, scores.data, target.view(-1).data)

        return loss, stats
```

**Comment by JianyuZhan** [comment]
Okay, so by default the `label_smoothing` behavior is off?  We would provide a train option for this or not? Besides, it would be better to document in code for this, at least reference is ok.

**Comment by helson73** [comment]
@JianyuZhan 
I created a new PR.
Here is the reference paper. Please see chapter 7.
https://arxiv.org/abs/1512.00567

**Comment by helson73** [comment]
About hyperparameters, I use followings, which is very close to the base setting in Tensor2Tensor.

`python -u train.py -gpuid 0 -save_model models/ja-en-1M/ja-en -data database/ja-en -layers 6 -rnn_size 512 -word_vec_size 512 -epochs 30 -max_grad_norm 0 -optim adam -encoder_type transformer -decoder_type transformer -position_encoding -dropout 0.1 -param_init 0 -warmup_steps 12000 -learning_rate 0.2 -decay_method noam -label_smoothing 0.1 -adam_beta2 0.98 -batch_size 80 -start_decay_at 31`

start_decay_at option is enabled in default, which should be turned off otherwise learning rate might be unnecessarily decreased. This should be fixed in the future, make that disabled when noam decay is enabled. 
On the other hand, warmup_steps option is varied case by case, which depends on batch size and training data size. If it is set properly, model will converge fast, and vice versa. One may want to check at which step learning rate should reach the peek.

**Comment by srush** [comment]
Going to close this one as it is basically all done. Huge props to @helson73 . I will add you to the README.
